eo-maven-plugin/src/main/java/org/eolang/maven/ParseMojo.java

```diff
@@ -119,46 +127,99 @@ public void exec() throws IOException {
+            this.targetDir.toPath().resolve(ParseMojo.DIR),
+            TranspileMojo.EXT
         );
-        new Save(
-            new XMLDocument(
+        if (ParseMojo.versioned(ver) && cached.exists()) {
+            Logger.info(
+                this,
+                "Found parsed in cache %s:%s",
+                Save.rel(source),
+                ver
+            );
+            new Save(
+                cached.content(),
+                target
+            ).save();
+        } else {
+            try {
+                new Syntax(
+                    name,
+                    new InputOf(source),
+                    new OutputTo(baos)
+                ).parse();
+                // @checkstyle IllegalCatchCheck (1 line)
+            } catch (final RuntimeException ex) {
+                if (this.failOnError) {
+                    throw new IllegalArgumentException(
+                        String.format("Failed to parse %s", source),
+                        ex
+                    );
+                }
+                Logger.warn(
+                    this, "Parse was skipped due to failOnError=false. In file %s with error: %s",
+                    source.toString(),
+                    ex.getMessage()
+                );
+                return;
+            }
+            final String content = new XMLDocument(
                 new Xembler(
                     new Directives().xpath("/program").attr(
                         "source", source.toAbsolutePath()
                     )
                 ).applyQuietly(new XMLDocument(baos.toByteArray()).node())
-            ).toString(),
-            target
-        ).save();
-        Logger.debug(
-            this, "Parsed %s to %s",
-            Save.rel(source), Save.rel(target)
-        );
+            ).toString();
+            new Save(content, target).save();
+            if (SafeMojo.versioned(ver)) {
+                cached.save(content);
+            }
+            Logger.debug(
+                this, "Parsed %s to %s",
+                Save.rel(source), Save.rel(target)
+            );
+        }
         tojo.set(AssembleMojo.ATTR_XMIR, target.toAbsolutePath().toString());
     }
 
+    /**
+     * Safely extract version attribute.
+     * @param tojo Source tojo
+     * @return Version value or empty string if attribute doesn't exist.
+     */
+    private static String verSafe(final Tojo tojo) {
+        String ver = ParseMojo.ZERO;
+        if (tojo.exists(AssembleMojo.ATTR_VERSION)) {
+            ver = tojo.get(AssembleMojo.ATTR_VERSION);
+        }
+        return ver;
+    }
+
+    /**
+     * Get hash for the version.
+     * @param ver Version to tag
+     * @return Version tag
+     */
+    private static String safeHash(final String ver) {
+        String hash;
+        try {
+            hash = new HashOfTag(ver).shortHash();
+        } catch (final IllegalArgumentException ex) {
```

Comment:
@mximp it's a very bad idea in general, to catch unchecked exception not in the global try/catch block

Reply:
@yegor256 I've created a puzzle for it.